src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -0,0 +1,170 @@
+                    {
+                        return true;
+                    }
+                }
+                else if (tokenType == JsonTokenType.StartArray)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    HandleStartArray(options, returnType, ref current, ref previous, ref arrayIndex);
+                }
+                else if (tokenType == JsonTokenType.EndArray)
+                {
+                    if (HandleEndArray(ref current, ref previous, ref arrayIndex))
+                    {
+                        return true;
+                    }
+                }
+                else if (tokenType == JsonTokenType.True)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    if (SetValueAsPrimitive(ref current, true))
+                    {
+                        return true;
+                    }
+
+                }
+                else if (tokenType == JsonTokenType.False)
+                {
+                    CheckForNonObjectReturn(ref current, options, returnType);
+                    if (SetValueAsPrimitive(ref current, false))
+                    {
+                        return true;
+                    }
+                }
+                else
+                {
+                    // todo: use boxed object with converter
+                    throw new InvalidOperationException();
+                }
+            }
+
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool SetValueAsPrimitive<T>(ref JsonObjectState current, in T value)
+        {
+            if (current.IsEnumerable() || current.IsPropertyEnumerable())
+            {
+                JsonObjectState.SetReturnValue(ref current, value, current.IsPropertyEnumerable());
+                return false;
+            }
+
+            if (current.ReturnValue == null)
+            {
+                current.ReturnValue = value;
+                return true;
+            }
+
+            ((JsonPropertyInfo<T>)current.PropertyInfo).Set(current.ReturnValue, value);
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void CheckForNonObjectReturn(ref JsonObjectState current, JsonConverterSettings options, Type returnType)
+        {
+            if (current.ClassInfo == null)
+            {
+                Debug.Assert(current.PropertyInfo == null);
+                JsonClassInfo returnClassInfo = options.GetOrAddClass(returnType);
+                current.ClassInfo = returnClassInfo;
+                current.PropertyInfo = returnClassInfo.GetPolicyProperty();
+            }
+        }
+
+        private static void GetPreviousState(ref List<JsonObjectState> previous, ref JsonObjectState state, int index)
```

Comment:
Can we return the `JsonObjectState` rather than passing it by ref, esp given the method name is `Get...`?

Reply:
Not as-is. As an optimization, `previous` is null initially (for simple classes that don't have nested objects)